src/rawposix/src/fs_calls.rs

```diff
@@ -987,3 +989,616 @@ pub fn fcntl_syscall(
+        let kernel_fd = convert_fd_to_host(virtual_fd as u64, virtual_fd_cageid, cageid);
+        if kernel_fd == -1 {
+            return syscall_error(Errno::EFAULT, "readlinkat", "Invalid Cage ID");
+        } else if kernel_fd == -9 {
+            return syscall_error(Errno::EBADF, "readlinkat", "Bad File Descriptor");
+        }
+        // Convert the path from relative path (lind-wasm perspective) to real kernel path (host kernel
+        // perspective)
+        let relpath = normpath(convpath(&path), cageid);
+        let relative_path = relpath.to_str().unwrap();
+        let full_path = format!("{}{}", LIND_ROOT, relative_path);
+        let c_path = std::ffi::CString::new(full_path).unwrap();
+
+        libcret = unsafe {
+            libc::readlinkat(
+                kernel_fd,
+                c_path.as_ptr(),
+                libc_buf.as_mut_ptr() as *mut i8,
+                libc_buflen,
+            )
+        };
+    }
+
+    if libcret < 0 {
+        let errno = get_errno();
+        return handle_errno(errno, "readlinkat");
+    }
+
+    // Convert the result from readlink to a Rust string
+    let libcbuf_str = unsafe { CStr::from_ptr(libc_buf.as_ptr() as *const i8) }
+        .to_str()
+        .unwrap();
+
+    // Adjust the result to remove LIND_ROOT prefix if present
+    let new_root = format!("{}/", LIND_ROOT);
+    let final_result = libcbuf_str
+        .strip_prefix(&new_root)
+        .unwrap_or(libcbuf_str);
+
+    // Check the length and copy the appropriate amount of data to buf
+    let bytes_to_copy = std::cmp::min(buflen, final_result.len());
+    unsafe {
+        std::ptr::copy_nonoverlapping(final_result.as_ptr(), buf as *mut u8, bytes_to_copy);
+    }
+
+    bytes_to_copy as i32
+}
+
+//------------------------------------RENAME SYSCALL------------------------------------
+/*
+ *  `rename` renames a file, moving it between directories if required.
+ *  Reference: https://man7.org/linux/man-pages/man2/rename.2.html
+ *
+ *  ## Arguments:
+ *   - `oldpath`: Current path of the file.
+ *   - `newpath`: New path for the file.
+ *
+ *  ## Implementation Details:
+ *   - Both paths are converted from the RawPOSIX perspective to the host kernel perspective
+ *     using `sc_convert_path_to_host`, which handles the LIND_ROOT prefixing and path normalization.
+ *   - The underlying libc::rename() is called with both converted paths.
+ *   - This can move files across directories and rename them simultaneously.
+ *
+ *  ## Return Value:
+ *   - `0` on success.
+ *   - `-1` on failure, with `errno` set appropriately.
+ */
+pub fn rename_syscall(
+    cageid: u64,
+    oldpath_arg: u64,
+    oldpath_cageid: u64,
+    newpath_arg: u64,
+    newpath_cageid: u64,
+    arg5: u64,
+    arg6: u64,
+    arg7: u64,
+    arg8: u64,
+    arg9: u64,
+    arg10: u64,
+    arg11: u64,
+    arg12: u64,
+) -> i32 {
+    let oldpath = sc_convert_path_to_host(oldpath_arg, oldpath_cageid, cageid);
```

Comment:
Do we handle the case where this path conversion exceeds the filename, etc. limit?